File: Kanghyun/[BOJ] 괄호 .cpp
#include <iostream>
#include <stack>
#include <vector>
#include <string>
using namespace std;
int main(void) {
	int T;
	cin >> T;

	string brakets;
	vector<string> ans;

	for (int i = 0; i < T; i++) {
		stack<char> stk;
		cin >> brakets;
		bool success = true;
		for (int j = 0; j < brakets.length(); j++) {
			if (brakets[j] == '(')
				stk.push(brakets[j]);
			else
				if (!stk.empty() && stk.top() == '(')
					stk.pop();
				else {
					success = false;
					break;
				}
					

		}
		if (stk.empty() && success)
			ans.push_back("YES");
		else
			ans.push_back("NO");
	}

	for (int i = 0; i < T; i++) 
		cout << ans[i]<< "\n";
	
}

File: Parksujin/[BOJ] 괄호.py
T = int(input())
for i in range(T):
  res = []
  ga = input()
  for j in ga:
    if j == '(':
      res.append(j)
    else: # pop을 하는데 스택이 비면 no 출력
      if res: # 비지 않으면
        res.pop()
      else:
        print('NO')
        break
  else: # for-else문 : for문이 break로 끊기지 않았을때 실행
    if not res: # pop에 끊기지 않고 스택이 비면 yes 출력
      print('YES')
    else:
      print('NO')
          
        
      
   